contracts/ThreeHeapOrdering.sol

```diff
@@ -0,0 +1,306 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+library ThreeHeapOrdering {
+    struct Account {
+        address id; // The address of the account.
+        uint96 value; // The value of the account.
+    }
+
+    struct HeapArray {
+        Account[] accounts; // All the accounts.
+        uint256 size; // The size of the heap portion of the structure, should be less than accounts length, the rest is an unordered array.
+        mapping(address => uint256) indexes; // A mapping from an address to a index in accounts.
+    }
+
+    /// CONSTANTS ///
+
+    uint256 private constant ROOT = 0;
+
+    /// ERRORS ///
+
+    /// @notice Thrown when the address is zero at insertion.
+    error AddressIsZero();
+
+    /// INTERNAL ///
+
+    /// @notice Updates an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with value `_formerValue` or when `_id` is not in the `_heap` with `_formerValue` equal to 0.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to update.
+    /// @param _formerValue The former value of the account to update.
+    /// @param _newValue The new value of the account to update.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    function update(
+        HeapArray storage _heap,
+        address _id,
+        uint256 _formerValue,
+        uint256 _newValue,
+        uint256 _maxSortedUsers
+    ) internal {
+        uint96 formerValue = SafeCast.toUint96(_formerValue);
+        uint96 newValue = SafeCast.toUint96(_newValue);
+
+        uint256 size = _heap.size;
+        uint256 newSize = computeSize(size, _maxSortedUsers);
+        if (size != newSize) _heap.size = newSize;
+
+        if (formerValue != newValue) {
+            if (newValue == 0) remove(_heap, _id, formerValue);
+            else if (formerValue == 0) insert(_heap, _id, newValue, _maxSortedUsers);
+            else if (formerValue < newValue) increase(_heap, _id, newValue, _maxSortedUsers);
+            else decrease(_heap, _id, newValue);
+        }
+    }
```

Comment:
These could be extracted to a base abstract contract (e.g. BaseHeap) and other methods could be defined virtual inside it and overridden in this implementation contract

Reply:
It's perhaps a matter of another PR though (TDD ; code, test, refactor)